### PR TITLE
fixes logic in locale redirect

### PR DESCRIPTION
**Pull request by Adebesin-Cell**
# fixes logic in locale redirect

This PRs handles edge cases causes multiple redirections in site

Fixes https://github.com/EveripediaNetwork/issues/issues/2172

### CODE REVIEW

**Comment by Royal-lobster** [comment]
@Adebesin-Cell it still doing infinite redirect, try this out

```ts
export function middleware(req: NextRequest) {
  const isMaintenanceMode = process.env.MAINTENANCE_MODE === 'true';

  // Check for maintenance mode first
  if (
    isMaintenanceMode &&
    req.nextUrl.pathname !== '/maintenance' &&
    !req.nextUrl.pathname.includes('.')
  ) {
    const url = req.nextUrl.clone();
    url.pathname = '/maintenance';
    return NextResponse.redirect(url.toString(), { status: 302 });
  }

  const userLocale = req.headers
    .get('accept-language')?
    .split(',')[0]
    .split('-')[0];

  // Check if the requested pathname already has a supported locale
  const isLocaleAlreadySet = languageData.some((lang) =>
    req.nextUrl.pathname.startsWith(`/${lang.locale}/`)
  );

  // If the user's locale is already set in the pathname, do not redirect
  if (isLocaleAlreadySet) {
    return NextResponse.next();
  }

  // Redirect based on user's locale if not already in the pathname
  const isLocaleSupported = languageData.some(
    (lang) => lang.locale === userLocale
  );

  // Redirect to user's locale if supported and not already in the pathname
  if (!isLocaleAlreadySet && isLocaleSupported) {
    const url = req.nextUrl.clone();
    url.pathname = `/${userLocale}${req.nextUrl.pathname}`;
    return NextResponse.redirect(url.toString(), { status: 302 });
  }

  // Redirect to default locale if user's locale is not supported and not already in the pathname
  if (!isLocaleAlreadySet && !isLocaleSupported) {
    const defaultLocale =
      languageData.find((lang) => lang.default)?.locale ?? 'en';
    const url = req.nextUrl.clone();
    url.pathname = `/${defaultLocale}${req.nextUrl.pathname}`;
    return NextResponse.redirect(url.toString(), { status: 302 });
  }

  return NextResponse.next();
}
```

**Comment by Adebesin-Cell** [comment]
> @Adebesin-Cell it still doing infinite redirect, try this out
> 
> ```ts
> export function middleware(req: NextRequest) {
>   const isMaintenanceMode = process.env.MAINTENANCE_MODE === 'true';
> 
>   // Check for maintenance mode first
>   if (
>     isMaintenanceMode &&
>     req.nextUrl.pathname !== '/maintenance' &&
>     !req.nextUrl.pathname.includes('.')
>   ) {
>     const url = req.nextUrl.clone();
>     url.pathname = '/maintenance';
>     return NextResponse.redirect(url.toString(), { status: 302 });
>   }
> 
>   const userLocale = req.headers
>     .get('accept-language')?
>     .split(',')[0]
>     .split('-')[0];
> 
>   // Check if the requested pathname already has a supported locale
>   const isLocaleAlreadySet = languageData.some((lang) =>
>     req.nextUrl.pathname.startsWith(`/${lang.locale}/`)
>   );
> 
>   // If the user's locale is already set in the pathname, do not redirect
>   if (isLocaleAlreadySet) {
>     return NextResponse.next();
>   }
> 
>   // Redirect based on user's locale if not already in the pathname
>   const isLocaleSupported = languageData.some(
>     (lang) => lang.locale === userLocale
>   );
> 
>   // Redirect to user's locale if supported and not already in the pathname
>   if (!isLocaleAlreadySet && isLocaleSupported) {
>     const url = req.nextUrl.clone();
>     url.pathname = `/${userLocale}${req.nextUrl.pathname}`;
>     return NextResponse.redirect(url.toString(), { status: 302 });
>   }
> 
>   // Redirect to default locale if user's locale is not supported and not already in the pathname
>   if (!isLocaleAlreadySet && !isLocaleSupported) {
>     const defaultLocale =
>       languageData.find((lang) => lang.default)?.locale ?? 'en';
>     const url = req.nextUrl.clone();
>     url.pathname = `/${defaultLocale}${req.nextUrl.pathname}`;
>     return NextResponse.redirect(url.toString(), { status: 302 });
>   }
> 
>   return NextResponse.next();
> }
> ```

How did you replicate the multiple redirect?

**Comment by Royal-lobster** [comment]
You can change your browser language, search for korean and make it to top

CleanShot 2024-01-05 at 15 00 33@2x

**Comment by Adebesin-Cell** [comment]
It keeps redirecting, I removed the logic and it seems to work fine aside the fact that the url doesn't change if the locale isn't present for translation.

**Comment by Royal-lobster** [comment]
> It keeps redirecting, I removed the logic and it seems to work fine aside the fact that the url doesn't change if the locale isn't present for translation.

yeah, i think nextjs already has the feature to detect the language of user right ? https://nextjs.org/docs/pages/building-your-application/routing/internationalization i believe we don't need to touch middleware

**Comment by Adebesin-Cell** [comment]
> > It keeps redirecting, I removed the logic and it seems to work fine aside the fact that the url doesn't change if the locale isn't present for translation.
> 
> yeah, i think nextjs already has the feature to detect the language of user right ? https://nextjs.org/docs/pages/building-your-application/routing/internationalization i believe we don't need to touch middleware

Yep! they handle it.

**Comment by Royal-lobster** [comment]
> Yep! they handle it.

Cool, make the change and feel free to ping here back so i can help test

**Comment by Adebesin-Cell** [comment]
> > Yep! they handle it.
> 
> Cool, make the change and feel free to ping here back so i can help test

https://nextjs.org/docs/pages/building-your-application/routing/internationalization#leveraging-the-next_locale-cookie

I'm using this rn in the system during language switch. And the system works by giving language locale it detects it renders based on that. You can test and see if that works or needs improvement.

**Comment by Royal-lobster** [comment]
@Adebesin-Cell Now its not detecting my browser language, its showing english version with prefered korean set in settings

File: next-i18next.config.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ const path = require('path')
 module.exports = {
   i18n: {
     defaultLocale: 'en',
-    locales: ['en', 'kr', 'zh'],
+    locales: ['en', 'kr'],
```
Comment (Royal-lobster) [inline next-i18next.config.js:7]: I believe we should also add this to next.config.js

CleanShot 2024-01-08 at 14 03 30@2x

**Comment by Adebesin-Cell** [comment]
> @Adebesin-Cell Now its not detecting my browser language, its showing english version with prefered korean set in settings

I'll look into that.

File: next-i18next.config.js (hunk 1 above)
Comment (Adebesin-Cell) [reply next-i18next.config.js:7]: I think it was abstracted into a new file and imported.

**Comment by Royal-lobster** [comment]
Its wierd that [automatic locale detection](https://nextjs.org/docs/pages/building-your-application/routing/internationalization#automatic-locale-detection) is not working :/ 

But check out this code, I made it work with middleware
https://github.com/Royal-lobster/Nextjs-Locale-Routing

CleanShot 2024-01-09 at 12 53 36

I think you are missing this step, when you use middleware, not only your current page gets passed to it, but also assets such as favicons and stuff. if you don't stop redirecting them, it goes into loop

CleanShot 2024-01-09 at 12 53 17

**Comment by Adebesin-Cell** [comment]
> Its wierd that [automatic locale detection](https://nextjs.org/docs/pages/building-your-application/routing/internationalization#automatic-locale-detection) is not working :/
> 
> But check out this code, I made it work with middleware https://github.com/Royal-lobster/Nextjs-Locale-Routing
> 
> CleanShot 2024-01-09 at 12 53 36
> 
> I think you are missing this step, when you use middleware, not only your current page gets passed to it, but also assets such as favicons and stuff. if you don't stop redirecting them, it goes into loop
> 
> CleanShot 2024-01-09 at 12 53 17

I tried this similar approach, it keeps giving "TOO MANY REDIRECTS"

**Comment by Royal-lobster** [comment]
There is also another case, if user's preference is default we don't redirect cuz default is just / and it redirects in loop. pushed fix in the commit. 

CleanShot 2024-01-10 at 10 46 40

Another thing that you need to do in this pr is, change kr to ko. AcceptedLanguages header passes korean preference as ko. but we are using kr as symbol for korean. currently its redirecting to 404

**Comment by Adebesin-Cell** [comment]
> There is also another case, if user's preference is default we don't redirect cuz default is just / and it redirects in loop. pushed fix in the commit.
> 
> CleanShot 2024-01-10 at 10 46 40
> 
> Another thing that you need to do in this pr is, change kr to ko. AcceptedLanguages header passes korean preference as ko. but we are using kr as symbol for korean. currently its redirecting to 404

yeah, there's a function for that. ``revertToKr``

**Comment by Royal-lobster** [comment]
> yeah, there's a function for that. `revertToKr`

Oh. Didn't study that code, but i removed them from utils folder cuz we are not using them in middleware now

**Review by Royal-lobster — APPROVED** [review]

**Comment by Adebesin-Cell** [comment]
Prolly related to this merged, we have cloudfare issues on [https://iq.wiki/activity](https://iq.wiki/ko/acitivity).

**Pull request merged**